### ISSUE TITLE
Project Update

### ISSUE THREAD

**Issue by yhyu13** [yhyu13/AlphaGOZero-python-tensorflow#4]
Hi,

My name is Yohan YU, your developer of this repository. Right now, I'm able to run a volta instance on the cloud, let's see how it goes. I will keep updating in a week.

**Comment by Zeta36** [comment]
Perfect!! :).

**Comment by yhyu13** [comment]
HI,

The training result on 11/8/2017 is released. The training move prediction is expected, but it doesn't generalize well on validation dataset. Could someone review my code on batch normalization?

The way I use batch norm:

In Network.py
`
def train(...):
model.mode = 'train'
sess.run(...)
...

def test(...):
model.mode = 'test'
sess.run(...)
...

In alphagozero_resnet_model.py
`
def batch_norm(...):
    return tf.contrib.layer.batch_norm(...,training=model.mode=='train',)

def build_train_op():
    ....
    update = tf.train.GraphKeys.UPDATE_OP
    with control_dependence(update):
           train = opt.apply_grads(...)
`

**Comment by yhyu13** [comment]
Hi,

Since NVIDIA tensorflow container only support py2.7 at this moment, all training would be done under "py2.7" branch.

**Comment by Zeta36** [comment]
Training acc > 70%!! It would be a wonderful result if you are able to validate the model can generalize correctly near this value. Why don't you use a validation dataset while you are training so you don't let the test to the final step after the train has finished? That'd tell you insubstantially whether  the model is generalizating well or not.

**Comment by yhyu13** [comment]
@Zeta36 The model is evaluate at the end of every epoch of training. It does surprises me that the model doesn't generalize well. I'm sure there are logical bugs in my code.

**Comment by yhyu13** [comment]
Hi,

I found one bug so far. The feature extract method is different from what DeepMind describe.

Mine used to be each player's past 8 moves and their corresponding states. However, features should be each player's stones extracted from the board in the past 8 states.

Therefore, I would preprocess dataset again, train another version of supervised learning go agent.

I've figured out the coroutine implementation of MCTS. It's in debugging phase now.

Yours,
Yohan Yu

**Comment by yuanfengpang** [comment]
![1](https://user-images.githubusercontent.com/31877689/32593399-1c82c2ac-c56b-11e7-93b1-79a1ea93d03d.png)
![2](https://user-images.githubusercontent.com/31877689/32593464-5fef99de-c56b-11e7-8ce7-269cf98c41ac.png)

Maybe you can try the new architecture of ResNet :)
reference:「Identity Mappings in Deep Residual Networks」arXiv:1603.05027 2016

**Comment by yhyu13** [comment]
@yuanfengpang 

Thanks! That paper introduced a full pre-activation residual block that overcome overfitting on ResNet-164, ResNet-110, and even ResNet-1001 because (1) the skip connection is an identity function in comparison to the original one, which makes chain rule backprop simpler (2) BN first impose stronger regularization, which reduce overfitting.

**Comment by yhyu13** [comment]
Hi

It's your developer Yohan Yu. It seems like Nvidia docker (which P3 instance requires) is still only support python 2.7 legacy. This implies we can't utilize uvloop to program coroutine in python 2.7. Anyway, I am asking nvidia admin when they will sync their framework to python 3. Otherwise, we will just stay in supervised learning. 

Regards,
Yohan Yu

**Comment by yhyu13** [comment]
Hi,

@mokemokechicken 

I manage to refractory the [APV_MCTS_2.py](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/blob/master/model/APV_MCTS_2.py) by using @classmethod. I believe now it not only looks better but also run faster. 

But as I said earlier, the main reason for speed up comes from going into more illegal move so that the expensive expansion is not invoked. 

When comparing [APV_MCTS.py](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/blob/master/model/APV_MCTS.py#L281) and [APV_MCTS_2.py](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/blob/master/model/APV_MCTS_2.py), I found even though their hyperparameters are the same, the search result is quite different consistently, see [this line in second version](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/blob/7cf0d11be9b37cf45c189f14d6785f19d03b095d/model/APV_MCTS_2.py#L18) and [this line in the first version](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/blob/7cf0d11be9b37cf45c189f14d6785f19d03b095d/model/APV_MCTS.py#L17). The second version consistently expand less leaf nodes than the first version. The best guess for me is that some coruntines bypass the ```virtual_loss_do()``` and manage to select the same location for both player, thus create illegal move in early game. 

I also found adding ```dirichlet``` noise is expensive because the data structure of my MCTS node has no phyonic way to do it [see this line](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/blob/7cf0d11be9b37cf45c189f14d6785f19d03b095d/model/APV_MCTS.py#L281).

The result is pretty good, with Cython conversion into C extension library:

**4 core 1000expansion/1600search per move without NN evaluation**

```
+Cython version:
 +2017-11-12 19:02:05,842 [59823] INFO     __main__: Global epoch 0 start.
 +2017-11-12 19:02:08,178 [59823] DEBUG    model.APV_MCTS_2_C: Searched for 2.33579 seconds
 +2017-11-12 19:02:08,179 [59823] DEBUG    __main__:
 +None
 +2017-11-12 19:02:08,179 [59823] INFO     __main__: Self-Play Simulation Game #0: 2.337 seconds
```

**Comment by yhyu13** [comment]
@Zeta36 @yuanfengpang

Hi,

I just made an update in a new branch called ```selfplay```. Everything new is in that branch. Thanks to [reversi-alpha-zero](https://github.com/mokemokechicken/reversi-alpha-zero), I am making progress. The dynamic resign threshold is implemented. Self play pipeline is functional. 

But there is only one piece missing: **I can't create a tensorflow on the flay (like in Keras)**. TF either requires me to initialize two(or more) models in the same graph, or rename all variables (It should prevent me from loading a trained model effectively). 

To deal with that, I remembered a previous project where i have to restore a checkpoint from VGG. Since the tensorflow .ckpt file stores dictionary like variable name and variable value. If variables share some part of their names, I can extract them and restore in another model. 

None of them sounds like an good idea. What do you guys think? Should I switch to Keras because it stores model in HDF5 and .json

**Comment by yhyu13** [comment]
Hi

I've fixed the bug left by MuGo and pygtp because they requires implementation that know each's api.

For details, please take a look at [5a19859 ](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/commit/19a37bec7727cf3441e1c0f3bb9eb58f3e4ae7dc)

**Comment by yhyu13** [comment]
Hi,

Checkout the latest update on the branch [selfplay](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/tree/selfplay). I add the gtp support for a popular/modern Go GUI Sabaki. It looks pretty and runs smoothly. Follow the instructions in the README.md. Enjoy

**Comment by yuanfengpang** [comment]
@yhyu13 
Hi, thank you for the update.I think it is too late to switch to Keras. Maybe it should prevent you from loading a trained model effectively, but I think accelerate the play process is more important.

I am developing a go bot based on MuGo. As you know even the test acc > 50%, the MuGO is still very weak, and can not beat the traditional MCTS. 

Recomputing the AlphaGo Zero weights will take about 1700 years on commodity hardware, see for example:  http://computer-go.org/pipermail/computer-go/2017-October/010307.html.So I am wondering how far I can get with the commodity hardware (like GTX 1070)

May I ask why you started this repository :)

PS:  I thought GoGUI was the best but Sabaki is greater!!!

**Comment by yhyu13** [comment]
@yuanfengpang 

AWS P3 instance offers Volta GPU in commodity price. The highest configuration is 8 Volta + 64 CPU which costs only $24/h. So for 72 hours training, it only cost less than $2000 which is even cheaper than a laptop. When Amazon launched P3, I thought there is nothing can stop me. But this project is harder than I thought.

**Comment by yuanfengpang** [comment]
@yhyu13
Hi, About Nov 15th Supervised Learning result. ``The MSE is between the actual outcome z ∈ {− 1, + 1} and the neural network value v, scaled by a factor of 1/4  to the range of 0–1.`` Figure 3 on the paper of alpha go ZERO.  
It can not be over 1, did you calculate it wrong.

**Comment by yhyu13** [comment]
@yuanfengpang 

Thanks! This explains my confusion. 

_Sent from my Samsung SAMSUNG-SM-G935A using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by yhyu13** [comment]
Quick update: the batch norm variables are not loaded from the checkpoint. The model isn't perform correctly. 

_Sent from my Samsung SAMSUNG-SM-G935A using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by yhyu13** [comment]
Hi,

I figured out how to build another model in tensorflow. It is done by setting up another tf.Graph(). This update features fully functional self play pipeline. In this version, I only implement two network: one "candidate" and another "best model" because my computer doesn't have memory to build another 6 layer mini AlphaGo. 

Let's me know what else you want to see in this project.

Yours,
Hang Yu

**Comment by yuanfengpang** [comment]
http://www.igoshogi.net/ai_ryusei/01/en/
I am going to join  a Go AI contest in Tokyo. If you are interested, I can register yours too.

**Comment by yhyu13** [comment]
@yuanfengpang 

Wish your top success in Japan! I'm interested but I have to attend NIPS2017 in Los Angeles during that exact time.

I heard GAP is part of a fully convolutional neural net. I haven't digged deep into that part but I will take a close look on how it works better. Thanks!

**Comment by yhyu13** [comment]
@yuanfengpang 

Take a look at a update of renet model [GAP](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/blob/a4b163cd4392de61b1f5127b8a216dda7971eb8b/model/alphagozero_resnet_full_model.py#L84). Is it the idea you are talking about? And which paper has discussed the GAP? Thanks!

**Comment by yuanfengpang** [comment]
Network In Network(https://arxiv.org/abs/1312.4400),  this paper should help you :)

**Comment by yhyu13** [comment]
@yuanfengpang 

Thank! From paper:
>One advantage of global average pooling over the fully connected
layers is that it is more native to the convolution structure by enforcing correspondences between
feature maps and categories. Thus the feature maps can be easily interpreted as categories confidence
maps. Another advantage is that there is no parameter to optimize in the global average pooling
thus overfitting is avoided at this layer. Futhermore, global average pooling sums out the spatial
information, thus it is more robust to spatial translations of the input.

**Comment by fuzzthink** [comment]
@yhyu13 The "selfplay" branch don't seem to exists.

**Comment by yhyu13** [comment]
Hi, @fuzzthink 

It should be a closed issue by now, I've integrated the self play pipeline into the master branch already. See this merge [Selfplay#6](https://github.com/yhyu13/AlphaGOZero-python-tensorflow/commit/9f7b01c7b242a9026afeb1d971296325e5e4e03b)
